parlai/core/metrics.py

```diff
@@ -251,6 +251,144 @@ def value(self) -> float:
+        # NOTE: hinting can be cleaned up with "from __future__ import annotations" when
+        # we drop Python 3.6
+        if other is None:
+            return self
+        full_true_positives: TScalar = self._true_positives + other._true_positives
+        full_true_negatives: TScalar = self._true_negatives + other._true_negatives
+        full_false_positives: TScalar = self._false_positives + other._false_positives
+        full_false_negatives: TScalar = self._false_negatives + other._false_negatives
+
+        # always keep the same return type
+        return type(self)(
+            true_positives=full_true_positives,
+            true_negatives=full_true_negatives,
+            false_positives=full_false_positives,
+            false_negatives=full_false_negatives,
+        )
+
+
+class PrecisionMetric(ConfusionMatrixMetric):
+    def value(self) -> float:
+        if self._true_positives == 0:
+            return 0.0
+        else:
+            return self._true_positives / (self._true_positives + self._false_positives)
+
+
+class RecallMetric(ConfusionMatrixMetric):
+    def value(self) -> float:
+        if self._true_positives == 0:
+            return 0.0
+        else:
+            return self._true_positives / (self._true_positives + self._false_negatives)
+
+
+class ClassificationF1Metric(ConfusionMatrixMetric):
+    def value(self) -> float:
+        if self._true_positives == 0:
+            return 0.0
+        else:
+            return (
+                2
+                * self._true_positives
+                / (
+                    2 * self._true_positives
+                    + self._false_negatives
+                    + self._false_positives
+                )
+            )
+
+
+class ClassificationMetric(AverageMetric):
+    """
+    Class takes sample-wise confusion matrix and computes precision, recall, f1,
+    weighted_f1 for classification.
+    """
+
+    @staticmethod
+    def compute_many(
+        true_positives: TScalar = 0,
+        true_negatives: TScalar = 0,
+        false_positives: TScalar = 0,
+        false_negatives: TScalar = 0,
+    ) -> Tuple[PrecisionMetric, RecallMetric, ClassificationF1Metric]:
+        return (
+            PrecisionMetric(
+                true_positives, true_negatives, false_positives, false_negatives
+            ),
+            RecallMetric(
+                true_positives, true_negatives, false_positives, false_negatives
+            ),
+            ClassificationF1Metric(
+                true_positives, true_negatives, false_positives, false_negatives
+            ),
+        )
+
+    @staticmethod
+    def compute_metrics(
+        predictions: List[str], gold_labels: List[str], positive_class: str
+    ) -> Tuple[List[PrecisionMetric], List[RecallMetric], List[ClassificationF1Metric]]:
+        precisions = []
+        recalls = []
+        f1s = []
+        for predicted, gold_label in zip(predictions, gold_labels):
+            true_positives = (
```

Comment:
could make it slight tighter with:

`true_positives = int(predicted == positive_class and gold_label == positive_class)`